### PR TITLE
Handle exceptions coming in from glib and objc in a better way

### PR DESCRIPTION
**Pull request by Therzok**
This PR fixes up exception handling so we always and properly attach stacktraces under mono.

Also, fixes stacktraces for non-error glib events

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs [hunk 1]
```diff
@@ -287,19 +287,36 @@ static void LoggerMethod (IntPtr logDomainPtr, LogLevelFlags logLevel, IntPtr me
 				LoggingService.LogError ("Disabling glib logging for the rest of the session");
 		}
 
+		static string GetStacktraceIfNeeded (LogLevelFlags flags)
+		{
+			if (flags.HasFlag (LogLevelFlags.Error | LogLevelFlags.Critical))
+				return string.Empty;
```
Comment (mrward) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs]: Not sure I understand why we are not getting the stack trace for Error or Critical? Can we add a comment why it is not needed here?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs]: Added: https://github.com/mono/monodevelop/pull/7733/commits/5790889871c9add4be64ef2d26a34a3f6ac5c188

**Comment by slluis** [comment]
@monojenkins backport release-8.1

**Comment by Therzok** [comment]
Tests passed in the md-addins PR: https://devdiv.visualstudio.com/DevDiv/_release?releaseId=340644&_a=release-contribution-tab-ms.vss-test-web.test-result-in-release-management

**Comment by mrward** [comment]
@Therzok With VS Mac 8.1.0.2669, which includes these changes, the gtk critical error callstack seems to be reported twice in the IDE log file. Not a big problem but thought I would mention it.

```
Critical Gtk error
MonoDevelop.Ide.Gui.GLibLogging+CriticalGtkException: Gtk-Critical: void gtk_container_remove(GtkContainer *, GtkWidget *): assertion `GTK_IS_TOOLBAR (container) || widget->parent == GTK_WIDGET (container)' failed
Stack trace: 
  at MonoDevelop.Ide.Gui.GLibLogging.LoggerMethod (System.IntPtr logDomainPtr, MonoDevelop.Ide.Gui.GLibLogging+LogLevelFlags logLevel, System.IntPtr messagePtr) [0x00029] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs:259 
  at MonoDevelop.Ide.Gui.GLibLogging+Log+<>c.<.cctor>b__23_0 (System.IntPtr domain, MonoDevelop.Ide.Gui.GLibLogging+LogLevelFlags level, System.IntPtr message, MonoDevelop.Ide.Gui.GLibLogging+LogFunc user_data) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs:103 
  at Gtk.Container.gtk_container_remove (System.IntPtr , System.IntPtr ) [0x00000] in <40f46e55eb67454ab904dc9e5644a3a1>:0 
  at Gtk.Container.Remove (Gtk.Widget widget) [0x00022] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/gtk/generated/Container.cs:419 
  at TestGtkCriticalErrorAddin.CreateGtkErrorHandler.Run () [0x00000] in <c102f7b50a6a4af588ad179b5c38adce>:0 
  at MonoDevelop.Components.Commands.CommandHandler.Run (System.Object dataItem) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs:75 
  at MonoDevelop.Components.Commands.CommandHandler.InternalRun (System.Object dataItem) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs:39 
  at MonoDevelop.Components.Commands.CommandManager.DefaultDispatchCommand (MonoDevelop.Components.Commands.ActionCommand cmd, MonoDevelop.Components.Commands.CommandInfo info, System.Object dataItem, System.Object target, MonoDevelop.Components.Commands.CommandSource source) [0x0006e] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs:1637 
  at MonoDevelop.Components.Commands.CommandManager.DispatchCommand (System.Object commandId, System.Object dataItem, System.Object initialTarget, MonoDevelop.Components.Commands.CommandSource source, System.Nullable`1[T] time, MonoDevelop.Components.Commands.CommandInfo sourceUpdateInfo) [0x003b5] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs:1600 
  at MonoDevelop.Components.Commands.CommandManager.DispatchCommand (System.Object commandId, System.Object dataItem, System.Object initialTarget, MonoDevelop.Components.Commands.CommandSource source, MonoDevelop.Components.Commands.CommandInfo sourceUpdateInfo) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs:1473 
  at MonoDevelop.Components.Mac.MDMenuItem+<>c__DisplayClass14_0.<Run>b__0 () [0x0005c] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenuItem.cs:90 
  at GLib.Timeout+TimeoutProxy.HandlerInternal (System.IntPtr data) [0x00017] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/glib/Timeout.cs:38 
  at Gtk.Application.gtk_main () [0x00000] in <40f46e55eb67454ab904dc9e5644a3a1>:0 
  at Gtk.Application.Run () [0x00001] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/gtk/Application.cs:145 
  at MonoDevelop.Ide.IdeStartup.Run (MonoDevelop.Ide.MonoDevelopOptions options) [0x00455] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:221 
  at MonoDevelop.Ide.IdeStartup.Main (System.String[] args, MonoDevelop.Ide.Extensions.IdeCustomizer customizer) [0x0009e] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:729 
  at Xamarin.Startup.MainClass.Main (System.String[] args) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/md-addins/Xamarin.Startup/Main.cs:11 
  at MonoDevelop.Ide.Gui.GLibLogging.LoggerMethod (System.IntPtr logDomainPtr, MonoDevelop.Ide.Gui.GLibLogging+LogLevelFlags logLevel, System.IntPtr messagePtr) [0x0008d] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs:275 
  at MonoDevelop.Ide.Gui.GLibLogging+Log+<>c.<.cctor>b__23_0 (System.IntPtr domain, MonoDevelop.Ide.Gui.GLibLogging+LogLevelFlags level, System.IntPtr message, MonoDevelop.Ide.Gui.GLibLogging+LogFunc user_data) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/GLibLogging.cs:103 
  at Gtk.Container.gtk_container_remove (System.IntPtr , System.IntPtr ) [0x00000] in <40f46e55eb67454ab904dc9e5644a3a1>:0 
  at Gtk.Container.Remove (Gtk.Widget widget) [0x00022] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/gtk/generated/Container.cs:419 
  at TestGtkCriticalErrorAddin.CreateGtkErrorHandler.Run () [0x00000] in <c102f7b50a6a4af588ad179b5c38adce>:0 
  at MonoDevelop.Components.Commands.CommandHandler.Run (System.Object dataItem) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs:75 
  at MonoDevelop.Components.Commands.CommandHandler.InternalRun (System.Object dataItem) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs:39 
  at MonoDevelop.Components.Commands.CommandManager.DefaultDispatchCommand (MonoDevelop.Components.Commands.ActionCommand cmd, MonoDevelop.Components.Commands.CommandInfo info, System.Object dataItem, System.Object target, MonoDevelop.Components.Commands.CommandSource source) [0x0006e] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs:1637 
  at MonoDevelop.Components.Commands.CommandManager.DispatchCommand (System.Object commandId, System.Object dataItem, System.Object initialTarget, MonoDevelop.Components.Commands.CommandSource source, System.Nullable`1[T] time, MonoDevelop.Components.Commands.CommandInfo sourceUpdateInfo) [0x003b5] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs:1600 
  at MonoDevelop.Components.Commands.CommandManager.DispatchCommand (System.Object commandId, System.Object dataItem, System.Object initialTarget, MonoDevelop.Components.Commands.CommandSource source, MonoDevelop.Components.Commands.CommandInfo sourceUpdateInfo) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs:1473 
  at MonoDevelop.Components.Mac.MDMenuItem+<>c__DisplayClass14_0.<Run>b__0 () [0x0005c] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenuItem.cs:90 
  at GLib.Timeout+TimeoutProxy.HandlerInternal (System.IntPtr data) [0x00017] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/glib/Timeout.cs:38 
  at Gtk.Application.gtk_main () [0x00000] in <40f46e55eb67454ab904dc9e5644a3a1>:0 
  at Gtk.Application.Run () [0x00001] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/gtk/Application.cs:145 
  at MonoDevelop.Ide.IdeStartup.Run (MonoDevelop.Ide.MonoDevelopOptions options) [0x00455] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:221 
  at MonoDevelop.Ide.IdeStartup.Main (System.String[] args, MonoDevelop.Ide.Extensions.IdeCustomizer customizer) [0x0009e] in /Users/vsts/agent/2.150.3/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:729 
  at Xamarin.Startup.MainClass.Main (System.String[] args) [0x00000] in /Users/vsts/agent/2.150.3/work/1/s/md-addins/Xamarin.Startup/Main.cs:11 
```

**Comment by Therzok** [comment]
_huh_, that's interesting. I'll take a look

**Pull request merged**